src/ocean/util/container/map/TreeMap.d

```diff
@@ -0,0 +1,488 @@
+
+        Adds a new node to the map for key or obtains the existing if already in
+        the map.
+
+        Params:
+            key   = the key to add a new node for or obtain the existing one
+            added = outputs true a new node was added or false if key was found
+                    in the map so that the existing node is returned
+
+        Returns:
+            the node associated with key. If added is true then it was newly
+            created and added, otherwise it is the already existing node in the
+            tree.
+
+    ***************************************************************************/
+
+    public T* put ( ulong key, out bool added )
+    {
+        return &(this.put_(key, added).value);
+    }
+
+    /***********************************************************************
+
+        Looks up the value for key in the map.
+
+        Params:
+            key = the key to look up node for
+
+        Returns:
+            the value associated with key or null if not found.
+
+    ***********************************************************************/
+
+    private T* opIn_r ( ulong key )
+    {
+        return &((cast(Node*)eb64_lookup(&this.root, key)).value);
+    }
+
+    /***********************************************************************
+
+        Obtains the first value in the map.
+
+        Returns:
+            the first value or `null` if the map is empty.
+
+    ***********************************************************************/
+
+    public T* first ()
+    {
+        return this.getBoundary!(true)();
+    }
+
+    /***********************************************************************
+
+        Obtains the last value in the map.
+
+        Returns:
+            the last value or `null` if the map is empty.
+
+    ***********************************************************************/
+
+    public T* last ()
+    {
+        return this.getBoundary!(false)();
+    }
+
+    /***********************************************************************
+
+        foreach iterator over nodes in the tree. Any tree modification is
+        permitted during iteration.
+
+    ***********************************************************************/
+
+    public int opApply ( int delegate ( ref ulong key, ref T* node ) dg )
+    {
+        int stop = 0;
+
+        for (auto eb_node = eb64_first(&this.root); eb_node && !stop;)
+        {
+            // Backup node.next here because dg() may change or delete node!
+            auto next = eb_node.next;
+            auto node = cast(Node*)eb_node;
+            auto value_ptr = &node.value;
+            stop = dg(node.node.key, value_ptr);
+            eb_node = next;
+        }
+
+        return stop;
+    }
+
+
+    /***********************************************************************
+
+        Removes node from the map and deallocates it.
+
+        DO NOT USE node from this point!
```

Comment:
I guess this comment no longer applies.

Reply:
Yeah. It should be "DO NOT USE any previous pointers to the accompanying value from this point!" or something similar.